Implement powermod for negative exponents

anthonyclays opened the pull request:
As I suggested in #15350, allowing `powermod` to take negative exponents is both a natural and useful addition to the function, so I took a stab at implementing this.
Thoughts?

anthonyclays committed:
Implemented powermod for negative exponents

tkelman commented:
Looks like there are a few places in `test/numbers.jl` that need to be updated here as well.

anthonyclays commented:
I updated all other appearances of `powermod`. Perhaps there should be some more tests for BigInts as well?

tkelman commented on test/numbers.jl:
```diff
@@ -2359,8 +2359,8 @@ end
 @test_throws DivideError powermod(1,0,0)
 @test_throws DivideError powermod(1,0,big(0))
 # negative power domain error
```
comment's not so accurate any more

tkelman commented:
> Perhaps there should be some more tests for BigInts as well?

Sure! Improved test coverage is always welcome.

anthonyclays committed:
Removed negative exponent check for BigInts

GMP will automatically perform modular inversion followed by
exponentiation when it received a negative exponent, see
https://gmplib.org/manual/Integer-Exponentiation.html#Integer-Exponentiation

anthonyclays committed:
Additional tests for powermod with BigInt arguments

anthonyclays commented:
The Travis failure is due to a time limit exceeded. However, it still successfully ran the [`intfuncs` tests](https://travis-ci.org/JuliaLang/julia/jobs/121081627#L2115).

Because gmp has [built-in support](https://gmplib.org/manual/Integer-Exponentiation.html#Integer-Exponentiation) for negative exponents in the exact way I described above, I simply removed the negative number check from `powermod(::BigInt, ::BigInt, ::BigInt)` (1cd5ddc)

I also added a number of tests for powermod with BigInt arguments.

StefanKarpinski commented:
Thanks, @anthonyclays! The OS X CI job timed out so I've restarted it just to be thorough, but it seems unlikely that one will fail when all the rest have passed.

anthonyclays commented:
Travis timed out again.

yuyichao commented:
Restarting the travis doesn't make it run on a new (merge) commit so it still runs all the tests on OSX. Please rebase on current master and force push.

anthonyclays commented:
The x86_64 linux build failed on Travis:
```
From worker 4:	  Test threw an exception of type ReadOnlyMemoryError
From worker 4:	  Expression: rationalize(Int64,prevfloat(0.1)) == 1 // 10
From worker 4:	  ReadOnlyMemoryError()
```
This seems unrelated... ?

anthonyclays commented:
Third time's the charm for Travis, apparently. Can this be merged?
